### ISSUE TITLE
Deriving an API server by using a "HasHandler" class for each API endpoint

### ISSUE THREAD

**Issue by runeksvendsen** [haskell-servant/servant#619]
If we have an API like this:

```
type GetUser   = "users" :> Get '[JSON] [User]
type GetAlbert = "albert" :> Get '[JSON] User
type GetIsaac  = "isaac" :> Get '[JSON] User

type UserAPI2 = GetUser :<|> GetAlbert :<|> GetIsaac
```

we currently define a server for the API by referencing the position of the endpoint within the API type:

```
server2 :: Server UserAPI2
server2 = return users2
     :<|> return albert
     :<|> return isaac
```

How about using a `HasHandler` class instead, for each endpoint within the API, like this:

```
class HasHandler endpoint where
    handle :: Server endpoint

instance HasHandler GetUser where
    handle = return users2

instance HasHandler GetAlbert where
    handle = return albert

instance HasHandler GetIsaac where
    handle = return isaac
```

Now, `server2` can be automatically derived by servant, and if an API endpoint doesn't have a `HasHandler` instance, the user sees a nice error like "no instance for HasHandler ("albert" :> Get '[JSON] User)" rather than the often unintelligible output of eg. plugging in handlers in the wrong order, or missing a handler, when using the `:<|>` combinator.

Is there something I'm missing here, or would this be a superior way to define servers for an API?

I think we could also use a similar class for the client functions, to make it easy get the appropriate client functions, such that we here also reference endpoints within an API, rather than the position of the endpoints within the API. But I'm not sure exactly how this should be structured.

**Comment by soenkehahn** [comment]
I think this breaks composability. Imagine I want to use your `UserAPI2` in one of my `servant` apps:

``` haskell
type MyApi =
  "userapi" :> UserAPI2 :<|>
  SomethingElse
```

Now do I need to provide instances for `HasHandler ("userapi" :> GetUser)`, etc.? Or will it just somehow use the ones from above?

Also, what if I want to reuse an API, but provide a different server implementation?

I do think this is an interesting idea, but I guess there are lots of open questions. So I think -- if you're interested in working on this -- this should be done in a separate package on top of `servant`/`servant-server`.

**Comment by alpmestan** [comment]
I think [something like this](https://github.com/abbradar/servant-record) would be a better solution, for the reasons mentionned above by @soenkehahn.

**Comment by runeksvendsen** [comment]
Good points! I've now had the chance to consider this a bit more. 

Firstly, I must admit, I didn't even know I could combine an entire API using the `:>` combinator, like so:

```haskell
    type API = "subapi" :> (GetAlbert :<|> GetIsaac)
```

But, with regards to whether or not it would make sense to require separately specifying handlers for the above API (given that handlers for `GetAlbert` and `GetIsaac` already exist) I would say that, logically, this should be determined by whether or not we consider the type signature for an endpoint to be a unique identifier for that endpoint. I would say this is necessarily the case. And I think this is exactly what we want: we want an endpoint type signature to be a unique representation of an HTTP request, such that there can't be any confusion as to which handler receives an incoming request. In my opinion, this is the true quality of Servant: the ability to unambiguously describe an HTTP request in logical terms.

 In other words, is the following a valid API?

```haskell
type API = GetAlbert :<|> GetAlbert
```

It can't be, right? Because the two endpoints represent the same request that needs to be handled. No implementation can distinguish between those two endpoints, but would be forced to deliver the same request to two handlers (which may have distinct implementations). 

Therefore, I think it would make sense to require new `HasHandler` instances for `"subapi" :> GetAlbert` and `"subapi" :> GetIsaac`. Which could be as simple as

```haskell
import UserAPI.API
import UserAPI.Handlers ()

instance HasHandler ("subapi" :> GetAlbert) where
   handle = (handle :: Server GetAlbert)
instance HasHandler ("subapi" :> GetIsaac) where
   handle = (handle :: Server GetIsaac)

```
if you want to reuse the existing handler implementations.

With regards to reusing an API, but providing different handlers, this would be a simple matter of defining the API and `HasHandler` instances in separate modules. I think this should be standard practice anyway, and I would be surprised if it weren't already. But, as far as I can see, the fact that you cannot import a module without importing its type class instances, is solely a Haskell-related issue, which affects all projects. And the solution is always to define optional instances in modules separate from the type definition module.

Anyway, I appreciate discussing this with you guys, as you clearly know a lot more about this than I do. So please keep the critique coming if you think of more.

**Comment by runeksvendsen** [comment]
Further thoughts on the subject:

It's often very useful to be able to run different handlers on different physical VM instances, since sometimes you don't want the performance of requests to one endpoint to affect the performance of requests to another endpoint.

As far as I can see, this is not easily possible when defining a `Server` using the `:<|>` combinator. If, instead, handlers were defined using type class instances, it would be simple to develop, for example, a reverse proxy type, which could route requests on different endpoints to distinct hosts.

So, let's say we want to define a reverse proxy type for out `UserAPI2` API:

```haskell
type GetUser   = "users" :> Get '[JSON] [User]
type GetAlbert = "albert" :> Get '[JSON] User
type GetIsaac  = "isaac" :> Get '[JSON] User

type UserAPI2 = GetUser :<|> GetAlbert :<|> GetIsaac
```

Then, if we have `HasHandler` instances for all the endpoints, we could do something like:

```haskell
class HasHandler a => HasProxyTarget a where
   target :: BaseUrl
```

and then it would be possible to derive a reverse proxy server, provided that all endpoints have `HasProxyTarget` instances. And the handler implementations would easily be able to run on different VMs, as they are separate from the API which they serve.

I don't see how this would be possible (not elegantly anyway) if we want to re-use existing patterns, since the `Server` implementation of an API is currently tied to the entire API.

This would require more thought, of course, but it seems to me that something like this would be a lot easier to build on top of Servant if handlers were `HasHandler` instances of endpoints.

**Comment by alpmestan** [comment]
> In other words, is the following a valid API?
>
> ``` haskell
> type API = GetAlbert :<|> GetAlbert
> ```
>
> It can't be, right?

Well, servant accepts it. But what happens when you serve that API? The way servant-server interprets `:<|>` in this case is: there's no distinction between both endpoints that allows us to determine whether a request is headed for the first or second endpoint, so it tries them in order. If the request is a valid one, the first handler is always picked. If not, 404.

> Therefore, I think it would make sense to require new HasHandler instances for "subapi" :> GetAlbert and "subapi" :> GetIsaac. Which could be as simple as [...]

Right, but this means everytime you move an endpoint in your API or anything like that, you have to define a new instance. This, among other things, is what is meant above by "breaking modularity".

> ``` haskell
> class HasHandler a => HasProxyTarget a where
>    target :: a -> BaseUrl
> ```

The problem I see with this is that it requires to provide the URLs/hosts at compile time, given that the endpoint spec doesn't contain any BaseUrl, you have to pull it out of thin air.

Another thing: `handle` is of type `HasHandler a => Server a`. `Server` however is not "injective", if `Server a` is the same as `Server b`, it *does not* mean that `a` is the same as `b`. Indeed:

``` haskell
type API1 = Post '[JSON] Int
type API2 = Get '[JSON] Int
```

These two API types are different, and are interpreted differently by all the servant interpretations (server, client, docs, etc). However, `Server API1` is the same as `Server API2`: `Handler Int` in both cases. This is certainly going to cause some problems, because in many (most?) cases, when just saying `handle`, GHC probably won't be able to trace back the type for which you need an `HasHandler` instance. Why? Because all it will know in those cases is `Server api`, and you can't deduce `api` from `Server api` as evidenced by my example above. One workaround is to add a `Proxy api` argument to your typeclass method, which makes things uglier.

To sum up my thoughts on this: I'm glad you're thinking hard about a way to put handlers together more easily that by making sure you `:<|>` them together in the same order that they appear in the API type. However, the current approach/code might to make things simpler in most cases, there probably is more work to be done to arrive at something that'll save you time and GHC errors. Note that you can experiment with this in a new package of yours, share it on hackage, even if it's not perfect. We would need a better solution for something that would be merged into the core servant packages, but in the meantime you can still package it up and share it =)

This is definitely not an easy problem by the way!
